### PR TITLE
Tag Semantic Release Action

### PR DESCRIPTION
**Pull request by shoukoo**
# Tag semantic release - easy way to tag your release 

## Parameters
- *version-type* (required, options: major, minor or patch)
- *github-token* (required)
- *debug* (optional, default: false)
- *dry-run* (optional, default: false)

```yaml
name: first
on:
  push:
    branches:
      - master
jobs:
  create:
    runs-on: [self-hosted, zendesk-stable]
    steps:
      - uses: zendesk/checkout@v2
      - uses: zendesk/ga/tag-semantic-release@v2
        with:
          github-token: ${{ secrets.ORG_GITHUB_TOKEN }}
          version-type: patch
```

### CODE REVIEW

File: tag-semantic-release/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  push:
+    branches:
+      - master
+
+jobs:
+  create:
+    runs-on: [self-hosted, zendesk-stable]
+    steps:
+      - uses: zendesk/checkout@v2
+      - uses: zendesk/ga/tag-semantic-release@shoukoo-tag
```
Comment (sfdye) [inline tag-semantic-release/README.md:20]: remove `@shoukoo-tag`?

File: tag-semantic-release/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Tag semantic release
+
```
Comment (sfdye) [inline tag-semantic-release/README.md:2]: maybe a short description of what this action does?

File: tag-semantic-release/action.yaml [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  github-token:
+    required: true
+  debug:
+    default: false
+runs:
+  using: "composite"
+  steps:
+    - run: |
+
+        echo "tsr v0.0.8"
```
Comment (sfdye) [inline tag-semantic-release/action.yaml:16]: what's this?

File: tag-semantic-release/README.md (hunk 1 above)
Comment (shoukoo) [reply tag-semantic-release/README.md]: Updated to v2

File: tag-semantic-release/action.yaml (hunk 3 above)
Comment (shoukoo) [reply tag-semantic-release/action.yaml]: This is my way of versioning gthis action!

File: tag-semantic-release/action.yaml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    value: ${{ steps.main.outputs.version }}
+runs:
+  using: "composite"
+  steps:
+    - run: |
+
+        echo "tsr v0.0.11"
+        export GITHUB_TOKEN=${{ inputs.github-token }}
+        git config --global user.name 'zendesk-ga'
+        git config --global user.email 'zendeak-ga@zendesk.com'
```
Comment (everops-john) [inline tag-semantic-release/action.yaml:25]: typo of zendesk-ga

File: tag-semantic-release/action.yaml [hunk 5]
```diff
@@ -0,0 +1,96 @@
+name: 'Tag semantic release'
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:1]:
```suggestion
name: Tag semantic release
```

File: tag-semantic-release/action.yaml [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+description: 'Tag semantic release for your default branch'
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:2]:
```suggestion
description: Tag semantic release for your default branch
```

File: tag-semantic-release/action.yaml [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+author: 'Andy Hsieh'
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:3]:
```suggestion
author: Andy Hsieh
```

File: tag-semantic-release/action.yaml [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+inputs:
+  version-type:
+    required: true
+    default: 'patch'
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:7]:
```suggestion
    default: patch
```

File: tag-semantic-release/action.yaml [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        export GITHUB_TOKEN=${{ inputs.github-token }}
+        git config --global user.name 'zendesk-ga'
+        git config --global user.email 'zendesk-ga@zendesk.com'
+        git config --global hub.protocol https
+
+        if [ ${{ inputs.debug }} = true ]; then
+          set -x
+        fi
+
+        if [[ ! -x "$(which hub)" ]]; then
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:32]:
```suggestion
        if [[ ! -x $(which hub) ]]; then
```

File: tag-semantic-release/action.yaml [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+          echo "Missing hub cli, install it here:"
+          echo "https://github.com/github/hub"
+          exit 1
+        fi
+
+        if [[ ! -x "$(which git)" ]]; then
+          echo "Missing git command"
+          exit 1
+        fi
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:41]: I personally think it's pretty safe to say you can leave this check out.

File: tag-semantic-release/action.yaml [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+        # Last 5 recent tags
+        git fetch --tags --quiet
+        tags=$(git for-each-ref refs/tags --sort=-creatordate '--format=%(refname:lstrip=2)' --count=5)
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:45]:
```suggestion
        tags=$(git for-each-ref refs/tags --sort=-creatordate --format="%(refname:lstrip=2)" --count=5)
```

File: tag-semantic-release/action.yaml [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+        # Find the latest semantic version
+        current_tag="v0.0.0"
+        for tag in $tags; do
+          if [[ $tag =~ ^v[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3} ]]; then
+            current_tag=$tag
+            echo "Found the latest semantic tag $tag"
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:52]:
```suggestion
            echo "Found the latest semantic tag as: $tag"
```

File: tag-semantic-release/action.yaml [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+            break
+          fi
+        done
+
+        # Turn string to an array
+        tag_array=($(echo "$current_tag" | tr '.' '\n'))
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:58]: If you change your regexp above to:

```
if [[ $tag =~ ^v([0-9]{1,3})\.([0-9]{1,3})\.([0-9]{1,3}) ]]; then
```

Then you'll have the three components ready to go in:

```
${BASH_REMATCH[1]}
${BASH_REMATCH[2]}
${BASH_REMATCH[3]}
```

Also means you don't have to slice off the leading `v`.

File: tag-semantic-release/action.yaml [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+        if [[ $( echo "${#tag_array[@]}") != 3 ]]; then
+          echo "Invalid tag $current_tag"
+          exit 1
+        fi
+
+        type=${{ inputs.version-type }}
+        major=$(echo "${tag_array[0]}" | tr -d "v" )
+        minor=${tag_array[1]}
+        patch=${tag_array[2]}
+
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:70]:
```suggestion
```

File: tag-semantic-release/action.yaml [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+            patch=$(( $patch + 1 ))
+            ;;
+          *)
+            echo "Invalid version type: $type"
+            exit 1
+            ;;
+        esac
+
+        new_tag="v$major.$minor.$patch"
+        echo "::set-output name=version::$(echo $new_tag)"
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:89]:
```suggestion
        echo "::set-output name=version::$new_tag"
```

File: tag-semantic-release/action.yaml [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+        echo "New tag: $new_tag"
+
+        if [ ${{ inputs.dry-run }} = false ]; then
+          hub release create -m "$new_tag" $new_tag
+        fi
+      shell: bash
+      id: main
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:96]: Can drop these two lines. `bash` is default shell, and you're not using the `id`.

```suggestion
```

File: tag-semantic-release/action.yaml [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+            echo "Found the latest semantic tag $tag"
+            break
+          fi
+        done
+
+        # Turn string to an array
+        tag_array=($(echo "$current_tag" | tr '.' '\n'))
+
+        # Must have 3 elements in the array
+        if [[ $( echo "${#tag_array[@]}") != 3 ]]; then
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:61]:
```suggestion
        if [[ ${#tag_array[@]} != 3 ]]; then
```

File: tag-semantic-release/action.yaml [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+outputs:
+  version:
+    value: ${{ steps.main.outputs.version }}
+runs:
+  using: "composite"
+  steps:
+    - run: |
+
+        echo "tsr v0.0.11"
+        export GITHUB_TOKEN=${{ inputs.github-token }}
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:23]: This should be optional no, does the default token not pass through?

File: tag-semantic-release/README.md [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+jobs:
+  create:
+    runs-on: [self-hosted, zendesk-stable]
+    steps:
+      - uses: zendesk/checkout@v2
+      - uses: zendesk/ga/tag-semantic-release@v2
+        id: tag
+        with:
+          github-token: ${{ secrets.ORG_GITHUB_TOKEN }}
```
Comment (magnetikonline) [inline tag-semantic-release/README.md:26]: I would change your example to this:

```suggestion
          github-token: ${{ secrets.GITHUB_TOKEN }}
```

No need to use org token credits for our automation user, if you're only tagging your own local repository.

File: tag-semantic-release/action.yaml [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+        git config --global user.email 'zendesk-ga@zendesk.com'
+        git config --global hub.protocol https
+
+        if [ ${{ inputs.debug }} = true ]; then
+          set -x
+        fi
+
+        if [[ ! -x "$(which hub)" ]]; then
+          echo "Missing hub cli, install it here:"
+          echo "https://github.com/github/hub"
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:34]: Use GitHub action style error messages

```suggestion
          echo "::error::Missing hub CLI, install it here:"
          echo "::error::https://github.com/github/hub"
```

File: tag-semantic-release/action.yaml [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+            break
+          fi
+        done
+
+        # Turn string to an array
+        tag_array=($(echo "$current_tag" | tr '.' '\n'))
+
+        # Must have 3 elements in the array
+        if [[ $( echo "${#tag_array[@]}") != 3 ]]; then
+          echo "Invalid tag $current_tag"
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:62]:
```suggestion
          echo "::error::Invalid tag $current_tag"
```

File: tag-semantic-release/action.yaml [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+            major=$(( $major + 1 ))
+            ;;
+          "minor")
+            minor=$(( $minor + 1 ))
+            ;;
+          "patch")
+            patch=$(( $patch + 1 ))
+            ;;
+          *)
+            echo "Invalid version type: $type"
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:83]:
```suggestion
            echo "::error::Invalid version type: $type"
```

File: tag-semantic-release/action.yaml (hunk 10 above)
Comment (shoukoo) [reply tag-semantic-release/action.yaml:41]: Thanks removed

File: tag-semantic-release/action.yaml (hunk 16 above)
Comment (shoukoo) [reply tag-semantic-release/action.yaml:95]: main is needed to generate the output, will remove `shell` 👍

File: tag-semantic-release/action.yaml (hunk 18 above)
Comment (shoukoo) [reply tag-semantic-release/action.yaml:23]: You are right, the normal GitHub token is enough to tag the releases.

File: tag-semantic-release/action.yaml (hunk 13 above)
Comment (shoukoo) [reply tag-semantic-release/action.yaml:53]: The above regex script didn't work for. It didn't turn the matched result into an array. 
I also prefer to have an extra line to explicitly turn the matched string into an array. 👍

File: tag-semantic-release/action.yaml (hunk 16 above)
Comment (shoukoo) [reply tag-semantic-release/action.yaml]: you need to have the shell 👍 
https://github.com/zendesk/andy-test/runs/991437440?check_suite_focus=true

File: tag-semantic-release/action.yaml [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+    required: true
+  debug:
+    default: false
+  dry-run:
+    default: false
+outputs:
+  version:
+    value: ${{ steps.main.outputs.version }}
+runs:
+  using: "composite"
```
Comment (magnetikonline) [inline tag-semantic-release/action.yaml:18]:
```suggestion
  using: composite
```

**Pull request merged**